halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/service/distributed/kubernetes/v2/ha/HaKubectlServiceProviderFactory.java

```diff
@@ -0,0 +1,220 @@
+    services.put(Type.MONITORING_DAEMON, monitoringDaemonService);
+
+    // Orca TODO
+    services.put(Type.ORCA, orcaService);
+
+    // Redis
+    if (!haServices.containsAll(ALLOWED_HA_SERVICES)) {
+      services.put(Type.REDIS, redisService);
+    }
+
+    // Rosco TODO
+    services.put(Type.ROSCO, roscoService);
+
+    return new MapBackedKubectlServiceProvider(services);
+  }
+
+  private void addClouddriverServices(Map<Type, KubernetesV2Service> services, List<Type> haServices) {
+    if (haServices.contains(Type.CLOUDDRIVER)) {
+      services.put(Type.REDIS_FOR_CLOUDDRIVER, redisMasterForClouddriverService);
+      services.put(Type.REDIS_SENTINEL_FOR_CLOUDDRIVER, redisSentinelForClouddriverService);
+      services.put(Type.REDIS_SLAVE_FOR_CLOUDDRIVER, redisSlaveForClouddriverService);
+      services.put(Type.CLOUDDRIVER_CACHING, clouddriverCachingService);
+      services.put(Type.CLOUDDRIVER_RO, clouddriverRoService);
+      services.put(Type.CLOUDDRIVER_RW, clouddriverRwService);
+    } else {
+      services.put(Type.CLOUDDRIVER, clouddriverService);
+    }
+  }
+
+  private void addGateServices(Map<Type, KubernetesV2Service> services, List<Type> haServices) {
+    String gateProfileContents = "";
+    if (haServices.contains(Type.GATE)) {
+      services.put(Type.REDIS_FOR_GATE, redisForGateService);
+      gateProfileContents = gateProfileContents.concat("\n"
+          + "redis:\n"
+          + "  connection: ${services.redisForGate.baseUrl}\n");
+    }
+    if (haServices.contains(Type.CLOUDDRIVER)) {
+      gateProfileContents = gateProfileContents.concat("\n"
+          + "clouddriver:\n"
+          + "  baseUrl: ${services.clouddriveRo.baseUrl}\n");
+    }
+
+    if (gateProfileContents.isEmpty()) {
+      services.put(Type.GATE, gateService);
+    } else {
+      services.put(Type.GATE, kubernetesV2ServiceWrapperFactory.newSpringServiceInstance(gateService,
+          "ha",
+          gateProfileContents));
+    }
+  }
+
+  @Slf4j
+  private static class MapBackedKubectlServiceProvider extends KubectlServiceProvider {
+    private final Map<Type, KubernetesV2Service> services;
+
+    public MapBackedKubectlServiceProvider(Map<Type, KubernetesV2Service> services) {
+      this.services = new HashMap<>(services);
+    }
+
+    @Override
+    public List<SpinnakerService> getServices() {
+      return services.values().stream()
+          .map(s -> SpinnakerService.class.cast(s))
+          .collect(Collectors.toList());
+    }
+
+    @Override
+    public List<KubernetesV2Service> getServicesByPriority(List<SpinnakerService.Type> serviceTypes) {
+      List<KubernetesV2Service> result = new ArrayList<>();
+      for (SpinnakerService.Type type : serviceTypes) {
+        if (services.containsKey(type)) {
+          result.add(services.get(type));
+        }
+      }
+      result.sort((a, b) -> {
+        // Prioritize Redis services
+        if (a.getService().getArtifact() == SpinnakerArtifact.REDIS) {
```

Comment:
I would put the priority directly in the service rather than having logic here to resolve it, and then sort by the priority the services expose.